packages/zoe/src/contractSupport/ratio.js

```diff
@@ -0,0 +1,135 @@
+import './types';
+import { assert, details as X, q } from '@agoric/assert';
+import { Nat } from '@agoric/nat';
+import { natSafeMath } from '.';
+
+const { multiply, floorDivide } = natSafeMath;
+
+// make a Ratio, which represents a fraction. It is a pass-by-copy record.
+//
+// The natural syntax for the most common operations we want to support
+// are Amount * Ratio and Amount / Ratio. Since the operations want to adhere to
+// the ratio rather than the amount, we settled on a calling convention of
+// multiplyBy(Amount, Ratio) and divideBy(Amount, Ratio).
+//
+// The most common kind of Ratio can be applied to Amounts of a particular
+// brand, and produces results of the same brand. This represents a multiplier
+// that is only applicable to that brand. The less common kind of Ratio can be
+// applied to one particular brand of amounts, and produces results of another
+// particular brand. This represents some kind of exchange rate. The
+// brand-checking helps us ensure that normal Ratios aren't applied to amounts
+// of the wrong brand, and that exchange rates are only used in the appropriate
+// direction.
+
+const PERCENT = 100n;
+
+const ratioPropertyNames = ['numerator', 'denominator'];
+
+export function assertIsRatio(ratio) {
+  const propertyNames = Object.getOwnPropertyNames(ratio);
+  assert(
+    propertyNames.length === 2,
+    X`Ratio ${ratio} must be a record with 2 fields.`,
+  );
+  for (const name of propertyNames) {
+    assert(
+      ratioPropertyNames.includes(name),
+      X`Parameter must be a Ratio record, but ${ratio} has ${q(name)}`,
+    );
+  }
+  Nat(ratio.numerator.value);
+  Nat(ratio.denominator.value);
+}
+
+export function makeRatio(
+  numerator,
+  numeratorBrand,
+  denominator = PERCENT,
+  denominatorBrand = numeratorBrand,
+) {
+  assert(denominator > 0n, X`No infinite ratios!`);
+
+  return harden({
+    numerator: { value: Nat(numerator), brand: numeratorBrand },
+    denominator: { value: Nat(denominator), brand: denominatorBrand },
+  });
+}
+
+export function makeRatioFromAmounts(numeratorAmount, denominatorAmount) {
+  assert(denominatorAmount.value > 0, X`No infinite ratios!`);
+
+  return harden({
+    numerator: {
+      value: Nat(numeratorAmount.value),
+      brand: numeratorAmount.brand,
+    },
+    denominator: {
+      value: Nat(denominatorAmount.value),
+      brand: denominatorAmount.brand,
+    },
+  });
+}
+
+export function multiplyBy(amount, ratio) {
+  assertIsRatio(ratio);
+  assert(
+    amount.brand === ratio.denominator.brand,
+    X`amount's brand ${amount.brand} must match ratio's denominator ${ratio.denominator.brand}`,
```

Comment:
should this also have the values in the details?

Reply:
I don't think so. I expect the brand would make it pretty clear where to look for the problem. The error message would have to look like `Amount's brand must match ratio's denominator. (372.23 "Moe") * (57/100) "ETH"/"ETH"`, and we have no standard way to print amounts or ratios,